TEP-0116: Referencing Finally Task Results in Pipeline Results (#746)
tektoncd/community: teps/0112-referencing-finally-task-results-in-pipeline-results.md

marked as a `proposed` is a working document and subject to change. You can
denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing TEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

If there are new details that belong in the TEP, edit the TEP. Once a
feature has become "implemented", major changes should get new TEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/teps/tools/tep-template.md.template).

-->

<!--
This is the title of your TEP. Keep it short, simple, and descriptive. A good
title can help communicate what the TEP is and should be considered as part of
any review.
-->

<!--
A table of contents is helpful for quickly jumping to sections of a TEP and for
highlighting any additional information provided beyond the standard TEP
template.

Ensure the TOC is wrapped with
  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>

> tags, and then generate with `hack/update-toc.sh`.

Reviewer: please remove the guides for writing teps (the commented out sections)

- [Design Details](#design-details)
- [Design Evaluation](#design-evaluation)
  - [Reusability](#reusability)
  - [Simplicity](#simplicity)
  - [Flexibility](#flexibility)
  - [User Experience](#user-experience)
  - [Performance](#performance)
  - [Risks and Mitigations](#risks-and-mitigations)
  - [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Implementation Plan](#implementation-plan)
  - [Test Plan](#test-plan)
  - [Infrastructure Needed](#infrastructure-needed)
  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
  - [Implementation Pull Requests](#implementation-pull-requests)
- [References](#references)
<!-- /toc -->

## Summary

The `Finally` field was introduced to the `PipelineSpec` at the same level as `Tasks`, `Params`, and `Workspaces`. The finally field is a list of one or more `PipelineTasks` similar to `Tasks` that provide a guarantee that they are executed only once all `PipelineTasks` under `Tasks` have completed, regardless if that is because of success or error. This proposal builds on having the `Finally` field on the same level as `Tasks` and be able to reference `Finally` `Task Result`'s in `Pipeline` `Result`'s.

## Motivation

> Today, `PipelineTasks` in the `Tasks` field can propagate `Results` to `Pipelines` while `PipelineTasks` in the `Finally` cannot propagate `Results` to `Pipelines`. This work was left out of scope when we added support for `Finally` - see [issue](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#cannot-configure-pipeline-result-with-finally).

Author: done

-->

1. Being able to reference `Pipeline` `Result` in another `Pipeline` (thats for [Pipeline in Pipelines](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md))

### Use Cases

<!--
Describe the concrete improvement specific groups of users will see if the
Motivations in this doc result in a fix or feature.

Consider the user's:
- [role][role] - are they a Task author? Catalog Task user? Cluster Admin? e.t.c.
- experience - what workflows or actions are enhanced if this problem is solved?

[role]: https://github.com/tektoncd/community/blob/main/user-profiles.md
-->

Most use cases either deal with cleaning up a workspace or with notifications. 

Let's say for example we have a `Pipeline` that clones a github repo. This `Pipeline` emits the initialization key, as a `Results` through the `Finally` field, once the repo has been cloned. Let’s say the user wants to instantaneously clean up the workspace as soon as the repo has been cloned. The user would need the initialization key in order to clean up the repo, but would be unable to do so because they won’t be able to refer to the `Finally` `Task` `Result` in the `Pipeline Result`. 

> A workflow that is enhaced for the user is the ability to pass more data within `Pipelines`. This decreases the verbosity of having to create multiple `Tasks` within `Pipelines`, rather allowing users to pass around variables when need be. 

Reviewer 2: `s/enhaced/enhanced`

Author: done

Supporting passing Finally Tasks Results to Pipeline Results is particularly necessary now that we no longer embed the full TaskRuns’ status in PipelineRuns - as proposed in [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md). 

### Requirements

<!--
Describe constraints on the solution that must be met, such as:
- which performance characteristics that must be met?
- which specific edge cases that must be handled?
- which user scenarios that will be affected and must be accommodated?
-->

> A case that must be handled is making sure `Finally` `Task` `Results` are not referenced within a `Pipeline` `Task` field. Or in other words, can't be referenced in non-`Finally` fields.

Author: done

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. The "Design Details" section below is for the real
nitty-gritty.
-->

> We propose interpolating all the `Tasks` and `TaskRefs` within the `Task` field and `Finally` field before anything else. With this approach, we are able to differentiate all the references before storage.

Reviewer 2: Same, I am not sure I understand that sentence

> We propose interpolating all the `Tasks` and `TaskRefs` within the `Task` field and `Finally` field before anything else. With this approach, we are able to differentiate all the references before storage.

Author: we will place `Result Value`in different buckets according to what they are referencing. So, `Result Value` that reference `tasks` (`$(tasks.<task-name>.result.<result-name>)`) will be in one bucket, while `Result Value`'s referencing `finally tasks` will be in another bucket.

It will follow the same naming conventions as referencing `Pipeline` `Tasks`: ```$(finally.<finally-task-name>.result.<result-name>)```.

The implementation will follow the previous implementation of passing [`Task` `Results`](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) around, however one validation that must be done is not allowing `Finally` `Task` `Results` in non-`Finally` fields.

### Notes and Caveats

<!--
(optional)

Go in to as much detail as necessary here.
- What are the caveats to the proposal?
- What are some important details that didn't come across above?
- What are the core concepts and how do they relate?
-->
1. Not allowing `Finally` `Task` `Results` in non-`Finally` fields

> ## Design Details

Reviewer: please remove any irrelevant sections of the TEP template

- How will it be tested in isolation vs with other components?

No need to outline all the test cases, just the general strategy. Anything
that would count as tricky in the implementation and anything particularly
challenging to test should be called out.

All code is expected to have adequate tests (eventually with coverage
expectations).
-->

### Infrastructure Needed

<!--
(optional)

Use this section if you need things from the project or working group.
Examples include a new subproject, repos requested, GitHub details.
Listing these here allows a working group to get the process for these
resources started right away.
-->

### Upgrade and Migration Strategy

<!--
(optional)

Use this section to detail whether this feature needs an upgrade or
migration strategy. This is especially useful when we modify a
behavior or add a feature that may replace and deprecate a current one.
-->

### Implementation Pull Requests

<!--
Once the TEP is ready to be marked as implemented, list down all the GitHub
merged pull requests.

Note: This section is exclusively for merged pull requests for this TEP.
It will be a quick reference for those looking for implementation of this TEP.
-->

> ## References

Reviewer: please add references to the docs and issues here

Author: done

A workflow that is enhanced for the user is the ability to pass more data within `Pipelines`. This decreases the verbosity of having to create multiple `Tasks` within `Pipelines`, rather allowing users to pass around variables when need be. 

Supporting passing Finally Tasks Results to Pipeline Results is particularly necessary now that we no longer embed the full TaskRuns’ status in PipelineRuns - as proposed in [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md). 

### Requirements

A case that must be handled is making sure `Finally` `Task` `Results` are not referenced within a `Pipeline` `Task` field. Or in other words, can't be referenced in non-`Finally` fields. It should also not be supported in other `finally` `tasks` - i.e. finally task result references are only supported in `Pipeline` `Results`.

## Proposal

> We propose interpolating all the `Tasks` and `TaskRefs` within the `Task` field and `Finally` field before anything else. With this approach, we are able to differentiate all the references before storage. In other words, we will place `Result Value` in different buckets according to what they are referencing. So, `Result Value` that reference `tasks` (`$(tasks.<task-name>.result.<result-name>)`) will be in one bucket, while `Result Value`'s referencing `finally tasks` will be in another bucket.

Reviewer: ```suggestion
We propose supporting referencing `Results` from `Finally Tasks` in `Pipeline Results`.
```

References of `Results` from `finally` will follow the same naming conventions as referencing `Results` from `tasks`: ```$(finally.<finally-pipelinetask-name>.result.<result-name>)```. Note that `Results` from `tasks` have the format: ```$(tasks.<pipelinetask-name>.result.<result-name>)```.

> The implementation will follow the previous implementation of passing [`Task` `Results`](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) around, however one validation that must be done is not allowing `Finally` `Task` `Results` in `PipelineResults` and `finally tasks`.

Reviewer: ```suggestion
The implementation will follow the previous implementation of passing [`Task` `Results`](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) around, however one validation that must be done is not allowing `Finally` `Task` `Results` anywhere else except the `PipelineResults`.
```

Author: done

### Notes and Caveats

1. Not allowing `Finally` `Task` `Results` in `Finally` fields.

> 2. Not allowing `Finally` `Task` `Results` in `PipelineResults`.

Reviewer: thought this is what we're doing 🤔

Author: meant to say `PipelineTasks`

### Reusability

The implementation will follow the previous implementation of passing [`Task` `Results`](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) around.

## Alternatives

> The current behavior of referencing `Finally` `Tasks`, ```$(task.finally.taskname)``` is not explicit and can lead to future errors with the broadness of it. 

Reviewer: wait, is this supported?

Author: it is not, unsure where I saw that
